internal/util/function/rerank/model_function.go

```diff
@@ -22,293 +22,411 @@ import (
+		url = "https://api.siliconflow.cn/v1/rerank"
+	}
 
-	base, _ := url.Parse(endpoint)
-	base.Path = "/v2/rerank"
-	model := baseModel{
-		url:            base.String(),
-		maxBatch:       maxBatch,
-		queryKey:       "query",
-		docKey:         "documents",
-		truncateParams: truncateParams,
-		parseScores: func(body []byte) ([]float32, error) {
-			var rerankResp vllmRerankResponse
-			if err := json.Unmarshal(body, &rerankResp); err != nil {
-				return nil, fmt.Errorf("Rerank error, parsing vllm response failed: %v", err)
-			}
+	siliconflowClient, err := siliconflow.NewSiliconflowClient(apiKey)
+	if err != nil {
+		return nil, err
+	}
 
-			sort.Slice(rerankResp.Results, func(i, j int) bool {
-				return rerankResp.Results[i].Index < rerankResp.Results[j].Index
-			})
+	var modelName string
+	maxBatch := 128 // The document does not clearly state what the maximum batch size is.
+	rankParams := map[string]any{}
 
-			scores := make([]float32, 0, len(rerankResp.Results))
-			for _, result := range rerankResp.Results {
-				scores = append(scores, result.RelevanceScore)
+	for _, param := range params {
+		switch strings.ToLower(param.Key) {
+		case models.ModelNameParamKey:
+			modelName = param.Value
+		case models.MaxClientBatchSizeParamKey:
+			if maxBatch, err = parseMaxBatch(param.Value); err != nil {
+				return nil, err
+			}
+		case models.MaxChunksPerDocParamKey:
+			maxChunksPerDoc, err := strconv.Atoi(param.Value)
+			if err != nil {
+				return nil, fmt.Errorf("[%s param's value: %s] is not a valid number", models.MaxChunksPerDocParamKey, param.Value)
 			}
+			rankParams[models.MaxChunksPerDocParamKey] = maxChunksPerDoc
+		case models.OverlapTokensParamKey:
+			overlapTokens, err := strconv.Atoi(param.Value)
+			if err != nil {
+				return nil, fmt.Errorf("[%s param's value: %s] is not a valid number", models.OverlapTokensParamKey, param.Value)
+			}
+			rankParams[models.OverlapTokensParamKey] = overlapTokens
+		default:
+		}
+	}
+	if modelName == "" {
+		return nil, fmt.Errorf("siliconflow rerank model name is required")
+	}
 
-			return scores, nil
-		},
+	provider := siliconflowProvider{
+		baseProvider:      baseProvider{batchSize: maxBatch},
+		siliconflowClient: siliconflowClient,
+		url:               url,
+		modelName:         modelName,
+		params:            rankParams,
 	}
-	return &vllmProvider{baseModel: model}, nil
+	return &provider, nil
 }
 
-type teiProvider struct {
-	baseModel
+func (provider *siliconflowProvider) rerank(ctx context.Context, query string, docs []string) ([]float32, error) {
+	rerankResp, err := provider.siliconflowClient.Rerank(provider.url, provider.modelName, query, docs, nil, 30)
+	if err != nil {
+		return nil, err
+	}
+	scores := make([]float32, len(docs))
+	for i, result := range rerankResp.Results {
+		scores[i] = result.RelevanceScore
+	}
+	return scores, nil
 }
 
-type TEIResponse struct {
-	Index int     `json:"index"`
-	Score float32 `json:"score"`
+type cohereProvider struct {
```

Comment:
Why we squeezed all prociders in a single go file instead of spread them like embedding